omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java

```diff
@@ -231,10 +231,10 @@ public PatientIdentifier save(PatientIdentifier delegate) {
 		}
 		
 		if (needToAdd) {
+			service().savePatientIdentifier(delegate);
```

Comment:
Is there a reason why you do not start with `delegate.getPatient().addIdentifier(delegate)` before calling `service().savePatientIdentifier(delegate)`?

Reply:
Yes - This was exactly the problem.  calling `delegate.getPatient().addIdentifier(delegate)` adds this identifier to the `Set` of identifiers for this patient. At this point, this identifier has not been validated and therefore may not be valid. When `service().savePatientIdentifier(delegate)` is called, the advice triggers the validator and if the validator fails, an exception is thrown. Once this exception is thrown, hibernate still updates the `delegate.getPatient()` with the added identifier and at this point, it persists it in the db (remember it wasn't validated)

Reply:
The fix is therefore, call `service().savePatientIdentifier(delegate)` and let validation happen before OR move all this

```
if (delegate.isPreferred()) {
			for (PatientIdentifier pI : delegate.getPatient().getActiveIdentifiers()) {
				if (!pI.equals(delegate)) {
					pI.setPreferred(false);
				}
			}
		}
		
		boolean needToAdd = true;
		for (PatientIdentifier pi : delegate.getPatient().getActiveIdentifiers()) {
			if (pi.equals(delegate)) {
				needToAdd = false;
				break;
			}
		}
		
		if (needToAdd) {
			delegate.getPatient().addIdentifier(delegate);
		}
```

 logic to the openmrs-core service layer so that validation happens before these checks